adoc/chapters/programming_interface.adoc

```diff
@@ -17640,40 +17635,500 @@ member function template.
+The second is the swizzle operation defined by the member function.
+The indices used by the second swizzle are the indices produced by the first
+swizzle.
+For example, if the second swizzle references the first element, this means the
+element of the underlying [code]#vec# that corresponds to the first element
+produced by the first swizzle.
+
+a@
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+template <access::address_space AddressSpace, access::decorated IsDecorated>
+void load(size_t offset, multi_ptr<const DataT, AddressSpace, IsDecorated> ptr)
+----
+!====
+_Effects:_ Loads values from memory into elements of the underlying [code]#vec#
+object.
+A total of [code]#NumElements# values are loaded from memory, starting at the
+the address [code]#ptr + offset*sizeof(DataT)#.
+The first value from memory is written to the element in [code]#vec# that
+corresponds to the first element of the swizzle operation.
+The second value from memory is written to the element in [code]#vec# that
+corresponds to the second element of the swizzle operation, etc.
+
+a@
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+DataT& operator[](int index)
+const DataT& operator[](int index) const
+----
+!====
+_Returns:_ A reference to the element of the underlying [code]#vec# object that
+corresponds to the position [code]#index# of the swizzle operation.
+
+a@
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+template</*unspecified*/>
+__swizzled_vec__& operator=(const __swizzled_vec__</*unspecified*/>& rhs)
+----
+!====
+_Constraints:_ Available only when the element data type of [code]#rhs# is the
+same as [code]#DataT# and when the number of elements in the [code]#rhs# view
+is equal to [code]#NumElements#.
+
+_Effects:_ Assigns elements from the [code]#rhs# [code]#+__swizzled_vec__+#
+view to elements of "this" [code]#+__swizzled_vec__+# view.
+The value from the [code]#rhs# corresponding to the first element of the
+[code]#rhs# swizzle operation is assigned to the element of the underlying
+[code]#vec# object that corresponds to the first element of the "this" swizzle
+operation, etc.
+
+The order in which the elements are assigned is unspecified.
+Therefore, if the "this" view has any repeated elements in the swizzle, those
+elements may receive unspecified values.
+
+_Returns:_ A reference to the "this" [code]#+__swizzled_vec__+# view.
+
+a@
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+__swizzled_vec__& operator=(const DataT& rhs)
+----
+!====
+_Effects:_ Assigns the value [code]#rhs# to those elements of the underlying
+[code]#vec# object that have corresponding elements in the
+[code]#+__swizzled_vec__+# view.
+Elements in the underlying [code]#vec# object that do not have elements in the
+[code]#+__swizzled_vec__+# view are not assigned.
+
+_Returns:_ A reference to the "this" [code]#+__swizzled_vec__+# view.
+
+a@
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+__swizzled_vec__& operator=(const vec<DataT, NumElements>& rhs)
```

Comment:
Could this overload be combined with the previous one with `__swizzled_vec__` like we have for the later operators?

Reply:
I think it would end up being more verbose that way.  We can't use the simple "behaves as though the swizzle operation was first evaluated into a temporary" because these are assignment operators (see my other response).  Therefore, I think we would essentially end up with two descriptions anyways.

Since the third assignment operator (`operator=(const DataT&)`) also has its own description, it seemed better to me to have separate table entries for all three.

Reply:
Yeah that makes sense, I agree we can use the same wording in all cases, the assignment operators do need to behave differently.

Though I wonder if we could have generic wording which applies to whether the parameters or return type is a `vec` or a `__swizzled_vec__`. I believe all binary operators where there there is an assignment, so the lhs parameter should not be treated as if being converted to a temporary `vec`, and the lhs parameter is a `__swizzled_vec__` then the return type is a `__swizzled_vec__` rather than a `vec`.

So I was thinking we could have something to the effect of:
* Any rhs parameter that is a `vec` is read element by element.
* Any rhs parameter that is a `__swizzled_vec__` is treated as if being converted to a temporary `vec`.
* If the return type is `vec` (non-assignment):
  * Any lhs parameter is treated as the rhs parameters are.
* If the return type is `__swizzled_vec__` (assignment):
  * Any lhs parameter that is a `vec` is assigned to element by element.
  * Any lhs parameter that is a `__swizzled_vec__` is assigned to element-by-element according to the swizzle index transformation.

This may be a little more complex, but we could potentially combine the wording of all binary operators.

Reply:
I think your goal is to reduce the size of the spec by combining all of the non-assignment "operatorOP" functions into a single synopsis with a common description.  Although I agree that it's good to have a shorter spec, I think this particular change would not be an improvement.  I see two problems:

* The "operatorOP" functions don't all have the same signatures.  I've already combined them as much as possible in the cases where they do have a common signature.  In order to combine them further, we'd have to somehow indicate which overloads applied to which OPs.  I think that would end up making the spec less clear.

* I'm concerned that the using general wording like what you propose above will either end up being less precise or it will be very long and complicated.  Either way, I think this will not improve the clarity of the spec.

Are you opposed to keeping the current structure?

Reply:
I think that's a fair concern, I was hoping that we could avoid having too much repeated wording, but I agree the alternative is to have a very long and complex general wording, so we may be exchanging verbosity for complexity, which now that I think about it some more is not necessarily helping as much as I thought.

We definitely can't combine all of the operators because as you described some have different signatures, so there would need to be some repeated wording either way.

I'm happy with keeping the wording structure as it is :)